cpp/arcticdb/codec/encoded_field.hpp

```diff
@@ -308,131 +238,220 @@ struct EncodedField {
         return {*this, true};
     }
 
-    EncodedBlockCollection values() const {
+    [[nodiscard]] EncodedBlockCollection values() const {
         return {*this, false};
     }
 
+    void validate() const {
+        size_t shapes_count = 0;
+        for(const auto& shape : shapes()) {
+            util::check(shape.is_shape_, "Expected shape to have is_shape_set");
+            util::check(shape.codecs_[0].codec_ != Codec::UNKNOWN, "Unknown shape codec");
+            ++shapes_count;
+        }
+        util::check(shapes_count == static_cast<size_t>(shapes_size()), "Shape size mismatch: {} != {}", shapes_count, shapes_size());
+
+        size_t values_count = 0;
+        for(const auto& value : values()) {
+            util::check(!value.is_shape_, "Value has is_shape set");
+            util::check(value.codec().codec_type() != Codec::UNKNOWN, "Unknown codec in block {}", values_count);
+            ++values_count;
+        }
+        util::check(values_count == static_cast<size_t>(values_size()), "Shape size mismatch: {} != {}", values_count, values_size());
+    }
+
     EncodedBlock *add_shapes() {
-        util::check(shapes_count_ == 0, "Expected single shapes block");
-        auto block = new(blocks() + items_count()) EncodedBlock{true};
+        auto block = new(blocks() + (shapes_count_ * 2)) EncodedBlock{true};
         ++shapes_count_;
         return block;
     }
 
-    int shapes_size() const {
+    [[nodiscard]] int shapes_size() const {
         return shapes_count_;
     }
 
-    int values_size() const {
+    [[nodiscard]] int values_size() const {
         return values_count_;
     }
 
     void set_sparse_map_bytes(uint32_t bytes) {
         sparse_map_bytes_ = bytes;
     }
 
-    EncodedBlock *add_values() {
-        auto block = new(static_cast<void *>(blocks() + shapes_count_ + values_count_)) EncodedBlock{false};
+    EncodedBlock *add_values(EncodingVersion encoding_version) {
+        const bool old_style = encoding_version == EncodingVersion::V1;
+        size_t pos;
+        if(!old_style || is_scalar())
+            pos = shapes_count_ + values_count_;
+        else
+            pos = (values_count_ * 2) + 1;
+
+        auto block = new(static_cast<void *>(blocks() + pos)) EncodedBlock{false};
         ++values_count_;
         return block;
     }
 
-    EncodedField *mutable_ndarray() {
-        type_ = EncodedFieldType::kNdarray;
+    EncodedFieldImpl *mutable_ndarray() {
+        type_ = EncodedFieldType::NDARRAY;
         return this;
     }
 
-    const EncodedField &ndarray() const {
+    [[nodiscard]] const EncodedFieldImpl &ndarray() const {
         return *this;
     }
 
-    bool has_ndarray() const {
-        return type_ == EncodedFieldType::kNdarray;
+    [[nodiscard]] bool has_ndarray() const {
+        return type_ == EncodedFieldType::NDARRAY;
     }
 
-    std::string DebugString() const {
-        return "";
+    [[nodiscard]] std::string DebugString() const {
```

Comment:
Why is this in PascalCase and not in snake_case?

Reply:
It's mirroring the method from protobuf